runtime/integration-tests/src/experimental/env.rs

```diff
@@ -0,0 +1,317 @@
+use std::{
+	collections::HashMap,
+	error::Error,
+	fmt::Debug,
+	sync::{Arc, Mutex},
+};
+
+use cfg_primitives::{AuraId, BlockNumber, Index};
+use codec::{Decode, Encode};
+use frame_support::traits::GenesisBuild;
+use frame_system::EventRecord;
+use fudge::{
+	digest::{DigestCreator, DigestProvider, FudgeAuraDigest, FudgeBabeDigest},
+	inherent::{
+		CreateInherentDataProviders, FudgeDummyInherentRelayParachain, FudgeInherentParaParachain,
+		FudgeInherentTimestamp,
+	},
+	primitives::{Chain, PoolState},
+	state::StateProvider,
+	ParachainBuilder, RelaychainBuilder, TWasmExecutor,
+};
+use lazy_static::lazy_static;
+use polkadot_core_primitives::Block as RelayBlock;
+use polkadot_parachain::primitives::Id as ParaId;
+use polkadot_primitives::runtime_api::ParachainHost;
+use sc_executor::{WasmExecutionMethod, WasmExecutor};
+use sc_service::{TFullClient, TaskManager};
+use sp_api::ApiExt;
+use sp_block_builder::BlockBuilder;
+use sp_consensus_babe::digests::CompatibleDigestItem;
+use sp_consensus_slots::SlotDuration;
+use sp_core::H256;
+use sp_runtime::{
+	generic::BlockId,
+	traits::{BlakeTwo256, Block, Extrinsic},
+	AccountId32, DigestItem, Storage,
+};
+use sp_transaction_pool::runtime_api::TaggedTransactionQueue;
+use tokio::runtime::Handle;
+
+use crate::utils::{logs, time::START_DATE};
+
+pub trait Config:
+	frame_system::Config<AccountId = AccountId32, BlockNumber = BlockNumber>
+	+ pallet_pool_system::Config<
+		CurrencyId = CurrencyId,
+		Balance = Balance,
+		PoolId = PoolId,
+		TrancheId = TrancheId,
+	> + pallet_balances::Config<Balance = Balance>
+	+ pallet_investments::Config<InvestmentId = TrancheCurrency, Amount = Balance>
+	+ pallet_pool_registry::Config<
+		CurrencyId = CurrencyId,
+		PoolId = PoolId,
+		Balance = Balance,
+		ModifyPool = pallet_pool_system::Pallet<Self>,
+		ModifyWriteOffPolicy = pallet_loans::Pallet<Self>,
+	> + pallet_permissions::Config<Role = Role, Scope = PermissionScope<PoolId, CurrencyId>>
+	+ pallet_loans::Config<
+		Balance = Balance,
+		PoolId = PoolId,
+		CollectionId = CollectionId,
+		ItemId = ItemId,
+	> + orml_tokens::Config<CurrencyId = CurrencyId, Balance = Balance>
+	+ orml_asset_registry::Config<
+		AssetId = CurrencyId,
+		CustomMetadata = CustomMetadata,
+		Balance = Balance,
+	> + pallet_uniques::Config<CollectionId = CollectionId, ItemId = ItemId>
+	+ pallet_timestamp::Config<Moment = Moment>
+	+ pallet_aura::Config<Moment = Moment>
+{
+	const KIND: RuntimeKind;
+}
+
+pub type Header = sp_runtime::generic::Header<BlockNumber, BlakeTwo256>;
+
+type RelayCidp = Box<
+	dyn CreateInherentDataProviders<
+		RelayBlock,
+		(),
+		InherentDataProviders = (
+			FudgeInherentTimestamp,
+			sp_consensus_babe::inherents::InherentDataProvider,
+			FudgeDummyInherentRelayParachain<Header>,
+		),
+	>,
+>;
+
+type RelayDp = Box<dyn DigestCreator<RelayBlock> + Send + Sync>;
+
+pub enum EventRange {
+	All,
+	One(BlockNumber),
+	Range(BlockNumber, BlockNumber),
+	Latest,
+}
+
+pub trait Environment {
```

Comment:
The environment wraps different `fudge::companion` structs to be used independently of the runtime